### PR TITLE
Shadow MMR theGraph

### LINKED ISSUE #423: Bridger and shadow next version checklist
To simplify the design of shadow, we are going to use theGraph as the indexer of Ethereum header MMR, instead of current indexing implementation in shadow. This could simplify the design and operation of shadow service, btw, it will also eliminate the necessary of operating geth node(for mmr import usage).

Here are the todos and checklist:

- [x] Deploy mmr subgraph for Ethereum mainnet using theGrpha's hosted servcie. cc @xiaoch05  https://api.thegraph.com/subgraphs/name/darwinia-network/ethereum-mmr
- [x] Deploy mmr subgraph for ropsten testnet using theGrpha's hosted servcie. cc @xiaoch05 https://api.thegraph.com/subgraphs/name/darwinia-network/ropsten-mmr
- [x] New library to generate mmr proof using theGraph api.  @fewensa @xiaoch05 
  - WIP PR: https://github.com/darwinia-network/shadow/pull/165   This is used for get proof nodes position from member_leaf_index and last_leaf_index
  - Using return value of above PR, to retrieve hash nodes from theGraph by their positions, and reassemble them into same proof data as legacy api
- [x] Replace MMR related api calls in bridger using new solution, and integration test. @fewensa 
- [x] Remove mmr related codes in new version of shadow. And deploy a new shadow service using new version, update/remove services. @fewensa 
- [ ] Archive the old version and documents for shadow. The mmr import features in deprecated shadow is still useful for validation and testing. @xiaoch05 

Related:
https://github.com/darwinia-network/darwinia-bridge-sol/pull/123
#400

### PR DESCRIPTION
**Pull request by fewensa**
Closes #423

### CODE REVIEW

**Comment by fewensa** [comment]
```

$ cargo test --package shadow-liketh --test test_shadow test_mmr_root -- --exact --nocapture


warning: `shadow-liketh` (lib) generated 5 warnings
    Finished test [unoptimized + debuginfo] target(s) in 44.82s
     Running tests\test_shadow.rs (target\debug\deps\test_shadow-c454ffec47bfbf0b.exe)

running 1 test
[138, 183, 56, 104, 109, 43, 215, 240, 240, 52, 3, 101, 238, 92, 35, 189, 239, 153, 197, 113, 209, 102, 38, 7, 16, 137, 3, 214, 70, 208, 179, 50]
test test_mmr_root ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.24s

```

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                    .map_err(|e| ShadowComponentError::MMR(format!("{:?}", e)))
+                {
+                    Ok(v) => Ok((item.id, v)),
+                    Err(e) => Err(e),
+                }
+            })
+            .collect::<Vec<ShadowComponentReuslt<(u64, [u8; 32])>>>()
+            .into_iter()
+            .collect::<ShadowComponentReuslt<Vec<(u64, [u8; 32])>>>()
+    }
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:132]: There are more clean ways to implement this? @AurevoirXavier

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            TheGraphResponse::NodeEntities(data) => Ok(data),
+        }
+    }
+
+    //
+    pub async fn mmr_root(&self, leaf_index: u64) -> ShadowComponentReuslt<[u8; 32]> {
+        let position = mmr_client::mmr::leaf_index_to_pos(leaf_index);
+        let peak_positions = mmr_client::mmr::get_peaks(position);
+
+        let mmr_positions = self.query_position(peak_positions).await?;
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:154]: renaming, `position` are different to concepts like 'node'

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            member_leaf_index,
+            last_leaf_index,
+            proof,
+        };
+        Ok(EthereumReceiptWithMMRProof { receipt, mmr_proof })
+    }
+}
+
+impl Shadow {
+    fn calc_peaks(
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:113]: bad naming

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            .await
+            .map_err(ShadowComponentError::from)?;
+
+        match response {
+            TheGraphResponse::NodeEntities(data) => Ok(data),
+        }
+    }
+
+    //
+    pub async fn mmr_root(&self, leaf_index: u64) -> ShadowComponentReuslt<[u8; 32]> {
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:150]: Can directly retrieve from on chain storage using optional parameter `at_block`

File: frame/assistants/shadow-liketh/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+serde-hex  = "0.1"
+serde-aux  = "0.6"
+arrayvec   = "0.7"
+
+codec      = { package = "parity-scale-codec", version = "2.3.1", features = ["derive", "full"] }
+
+gql_client = "1"
+
+reqwest    = { version = "0.11", features = ["rustls-tls", "json"] }
+mmr_client = { package = "merkle-mountain-range", git = "https://github.com/darwinia-network/shadow.git", branch = "add-mmr-client" }
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/Cargo.toml:36]: `merkle-mountain-range` is a independent third party lib, we use the customized branch because we modified it?

**Review by HackFisher — CHANGES_REQUESTED** [review]
Generally, we need more clean to the codes.

File: frame/assistants/shadow-liketh/src/shadow.rs (hunk 4 above)
Comment (HackFisher) [reply frame/assistants/shadow-liketh/src/shadow.rs:150]: better to directly query from block header, or return together with proof.

Ref: 
https://github.com/darwinia-network/darwinia-common/blob/17ce2d0d9668124bd707be92ff62c222aff99b25/frame/header-mmr/src/lib.rs#L164

File: frame/assistants/shadow-liketh/Cargo.toml (hunk 5 above)
Comment (fewensa) [reply frame/assistants/shadow-liketh/Cargo.toml:36]: Nope, the `merkle-mountain-range` package is defined by shadow
https://github.com/darwinia-network/shadow/blob/e60661c78078fedfccf26a799248d8e43b26dd7d/mmr-client/Cargo.toml#L2
but this folder name is `mmr-client` so I changed package's name to `mmr_client`

File: frame/assistants/shadow-liketh/src/shadow.rs (hunk 4 above)
Comment (HackFisher) [reply frame/assistants/shadow-liketh/src/shadow.rs:150]:
```
    /// get mmr root
    pub async fn get_mmr_root(&self, leaf_index: u32) -> ClientResult<subxt::sp_core::H256> {
        let block_number = leaf_index + 1;
        let header = self.header_by_number(block_number).await?;

        let mmr_root = if let Some(header) = header {
            // get digest_item from header
            let log = header
                .digest()
                .logs()
                .iter()
                .find(|&x| x.as_other().is_some());
            if let Some(digest_item) = log {
                // get mmr_root from log
                let parent_mmr_root = digest_item.as_other().unwrap().to_vec();
                let parent_mmr_root = &parent_mmr_root[4..];
                if parent_mmr_root.len() != 32 {
                    return Err(ClientError::WrongMmrRootInDarwiniaHeader(
                        array_bytes::bytes2hex("", &parent_mmr_root),
                        block_number,
                    ));
                }
                let mut mmr_root: [u8; 32] = [0; 32];
                mmr_root.copy_from_slice(parent_mmr_root);
                subxt::sp_core::H256(mmr_root)
            } else {
                return Err(ClientError::NoMmrRootInDarwiniaHeader(block_number));
            }
        } else {
            return Err(ClientError::FailedToFetchDarwiniaHeader(block_number));
        };
        Ok(mmr_root)
    }
```

File: frame/assistants/shadow-liketh/src/shadow.rs (hunk 1 above)
Comment (fewensa) [reply frame/assistants/shadow-liketh/src/shadow.rs]: https://docs.rs/serde-hex/latest/serde_hex/

```
#[derive(Debug,Serialize,Deserialize)]
struct Foo {
   #[serde(with = "SerHex::<StrictPfx>")]
   bar: [u8;32]
}
```
the serde-hex crate is support deserialize to a fixed array

Comment (fewensa) [reply frame/assistants/shadow-liketh/src/shadow.rs]: https://github.com/darwinia-network/bridger/pull/429/commits/e929f66c344bcecf9f19392dece7ffa8025ad522

**Comment by fewensa** [comment]
@HackFisher Please review this pull request. If that's ok. I'll start testing and making modifications to the darwinia-ethereum bridge

File: bridges/pangolin-ropsten/bridge/src/command/handler/parcel.rs [hunk 6]
```diff
@@ -16,10 +18,14 @@ pub async fn handle_parcel(opts: ParcelOpts, basic: BasicOptions) -> color_eyre:
         bridge_config.shadow,
         bridge_config.ethereum,
         bridge_config.web3,
+        BridgeName::PangolinRopsten,
     )?;
 
+    // Darwinia client
+    let client = PangolinClientComponent::component(bridge_config.darwinia).await?;
+    let mmr_root = client.get_mmr_root(block).await?;
```
Comment (HackFisher) [inline bridges/pangolin-ropsten/bridge/src/command/handler/parcel.rs:26]: I was wrong here that it should be the root of ethereum mmr. 

We need to revert the changes to calculate that for nodes.

Comment (fewensa) [reply bridges/pangolin-ropsten/bridge/src/command/handler/parcel.rs]: Fixed https://github.com/darwinia-network/bridger/pull/429/commits/cbe5ff5f639c6af7d6c4acf39afd17723bda1e8d

**Comment by HackFisher** [comment]
https://github.com/darwinia-network/bridger/blob/shadow-mmr/thegraph/mmr/src/mmr-init.json#L5-L16

@xiaoch05 How the initial data of ropsten generated? Are you using shadow import from the scratch? If in that case, did you reset/check the production repsten shadow to be the same with the init data?

Cause I found the initial proof part, which should also be peak positions, are different with this init data. I'm wondering the ropsten shadow get messed up due to long reorg.

```
  "mmr_proof": {
    "member_leaf_index": 12244902,
    "last_leaf_index": 12244902,
    "proof": [
      "36f3d834cbe12a5a20b063c432b88f5506bdce03b93fa3aa035a5d82fd50177c",
      "b10a06336827182396eabf37e835e57252cee94fd4b493787a76a9869026a65e",
      "742fb2ccd76238a81486ed4e9236c8980587e9cb0e4a97d35b1f4488dc42d748",
      "5d82e3028dfe10b766a065e68276fce28bb33eaf8a46ea684465d68249e04d9e",
      "5fcabe50b10a4da4059f7212efab393e62cfa968f8a3435c29e4568bf62327cf",
      "8136d50cf9026a46c6eb7213e205eed92e803c685ba2aaf7b6321e15bdd3e78f",
      "0734bb02593082ac7484e12475be76ffa2be60f9d8c2190a4bd3ac264b03a337",
      "c6a3abb0af037766e02ce191766dd0721ee868a0dcea2a7f677935ae8fe1409f",
      "2100feb8734a2fe7812dae3071aea9c6b3e3ecab7516158928e3e6fc5051f047",
      "c5062ae0eb5ce65104ea2b5128d785a819181198ae39c99bc39cc072bcb6bbe3",
      "3414d48d47c248c3a66253ff563a11815ec4566981786e9f8f54c5c525c51706",
      "52f4e7ad9cc9fb819741dc9b430270f4ca705aeaa65c64c880f73a9af3b95160",
      "0de56976d432ff5a16b3cee3a4d174aa57b3ef23963a55bfa97661722933e456",
      "32e0af5ddfcc5fb22b6fda3be047c17d90d5b2bf37ec9ca342efa0ae4b771d0a",
      "324b474afbc1b7f1420eb10dd20b5c9df00ac95a9c44430fdbf3e01c01799736"
    ]
```

File: frame/assistants/shadow-liketh/src/mmr.rs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    Ok(peaks.iter().fold(last, |prev, &next| merge(&prev, &next)))
+}
+
+// mmr root
+// 1. helper::get_peaks(leaf_pos);
+// 2. query hash of peaks from db
+// 3. bag_rhs_peaks
+
+// first call gen_proof_positions to get the positions proof needed
+pub fn gen_proof_positions(mut pos: u64, leaf_pos: u64) -> (Vec<u64>, Vec<u64>, u64) {
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/mmr.rs:30]: The implementation copy from https://github.com/darwinia-network/shadow/pull/165 is buggy.

In the test, the `proof_index `/ `merkle_proof_part` is empty. which indicate it is wrong.

We should use a more stable and production battle tested lib: https://github.com/nervosnetwork/merkle-mountain-range

**Comment by xiaoch05** [comment]
> 

@fewensa You are confusing the configuration of ropsten and mainnet. ropsten's latest block is not that high

**Comment by xiaoch05** [comment]
> https://github.com/darwinia-network/bridger/blob/shadow-mmr/thegraph/mmr/src/mmr-init.json#L5-L16
> 
> @xiaoch05 How the initial data of ropsten generated? Are you using shadow import from the scratch? If in that case, did you reset/check the production repsten shadow to be the same with the init data?
> 
> Cause I found the initial proof part from classic ropsten shadow api, which should also be peak positions, are different with this init data. I'm wondering the ropsten shadow get messed up due to long reorg.
> 
> ```
>   "mmr_proof": {
>     "member_leaf_index": 12244902,
>     "last_leaf_index": 12244902,
>     "proof": [
>       "36f3d834cbe12a5a20b063c432b88f5506bdce03b93fa3aa035a5d82fd50177c",
>       "b10a06336827182396eabf37e835e57252cee94fd4b493787a76a9869026a65e",
>       "742fb2ccd76238a81486ed4e9236c8980587e9cb0e4a97d35b1f4488dc42d748",
>       "5d82e3028dfe10b766a065e68276fce28bb33eaf8a46ea684465d68249e04d9e",
>       "5fcabe50b10a4da4059f7212efab393e62cfa968f8a3435c29e4568bf62327cf",
>       "8136d50cf9026a46c6eb7213e205eed92e803c685ba2aaf7b6321e15bdd3e78f",
>       "0734bb02593082ac7484e12475be76ffa2be60f9d8c2190a4bd3ac264b03a337",
>       "c6a3abb0af037766e02ce191766dd0721ee868a0dcea2a7f677935ae8fe1409f",
>       "2100feb8734a2fe7812dae3071aea9c6b3e3ecab7516158928e3e6fc5051f047",
>       "c5062ae0eb5ce65104ea2b5128d785a819181198ae39c99bc39cc072bcb6bbe3",
>       "3414d48d47c248c3a66253ff563a11815ec4566981786e9f8f54c5c525c51706",
>       "52f4e7ad9cc9fb819741dc9b430270f4ca705aeaa65c64c880f73a9af3b95160",
>       "0de56976d432ff5a16b3cee3a4d174aa57b3ef23963a55bfa97661722933e456",
>       "32e0af5ddfcc5fb22b6fda3be047c17d90d5b2bf37ec9ca342efa0ae4b771d0a",
>       "324b474afbc1b7f1420eb10dd20b5c9df00ac95a9c44430fdbf3e01c01799736"
>     ]
> ```

the data in shadow is not corrent.
https://ropsten.shadow.darwinia.network/ethereum/mmr_leaf/10920274
`{
    "mmr_leaf": "0x2fd63eb95c8059a46b5afedb903ef1639b9414e9d38f70b3dfb0f733b7ced14f"
}`
but actual the block hash is
https://ropsten.etherscan.io/block/10920274
0xfd652086d220d506ae5b7cb80fde97d2f3f7028d346cc7d9d384a83d3d638532

File: frame/assistants/shadow-liketh/src/mmr.rs (hunk 7 above)
Comment (xiaoch05) [reply frame/assistants/shadow-liketh/src/mmr.rs:30]: the `merkle_proof_part` is empty because the testcase use the same block_number as leaf_index to verify the number.
just like `./target/release/mmr https://api.thegraph.com/subgraphs/name/darwinia-network/ropsten-mmr 12244903 12244902`
if we use the next leaf_index to verify the block, it must not be empty, like
`./target/release/mmr https://api.thegraph.com/subgraphs/name/darwinia-network/ropsten-mmr 12244904 12244902`

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            .await
+            .map_err(ShadowComponentError::from)?;
+
+        match response {
+            TheGraphResponse::NodeEntities(data) => Ok(data),
+        }
+    }
+
+    pub async fn mmr_root(&self, leaf_index: u64) -> ShadowComponentReuslt<[u8; 32]> {
+        let position = ckb_merkle_mountain_range::leaf_index_to_mmr_size(leaf_index);
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:142]: This is right now.

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        Ok(mmr_root)
+    }
+
+    pub async fn mmr_proof(
+        &self,
+        tx_number: u64,
+        last_leaf: u64,
+    ) -> ShadowComponentReuslt<Vec<[u8; 32]>> {
+        let tx_position = ckb_merkle_mountain_range::leaf_index_to_pos(tx_number);
+        let leaf_pos = ckb_merkle_mountain_range::leaf_index_to_pos(last_leaf);
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:167]:
```suggestion
        let mmr_size = ckb_merkle_mountain_range::leaf_index_to_mmr_size(last_leaf);
```

**Comment by fewensa** [comment]
Don't care about subql deploy error. will be fix late

```
 subquery --token *** deployment deploy --org fewensa --key subql-bridge-pangolin --branch shadow-mmr --sub-folder subql/ethereum-v1/pangolin --indexer-image-version v0.28.2 --query-image-version v0.12.0 --type stage
Deploy subql/ethereum-v1/pangolin to subql-bridge-pangolin
[{"createdAt":"2022-03-31T10:25:45.454Z","updatedAt":"2022-05-17T03:48:30.933Z","id":12867,"projectKey":"fewensa/subql-bridge-pangolin","version":"77b1a25da3d357349f2fd40bc266006a6bc538d4","status":"stopped","cluster":"jm","indexerImage":"onfinality/subql-node:v0.31.1","queryImage":"onfinality/subql-query:v0.13.0","subFolder":"subql/ethereum-v1/pangolin","endpoint":null,"dictEndpoint":"","type":"primary","queryUrl":"https://api.subquery.network/sq/fewensa/subql-bridge-pangolin","configuration":{"config":{"query":{},"indexer":{"batchSize":30}},"role":"master"}},{"createdAt":"2022-04-19T11:07:03.806Z","updatedAt":"2022-05-17T03:48:42.610Z","id":13234,"projectKey":"fewensa/subql-bridge-pangolin","version":"6a5acf6ea31612a52595e9a5e393b2375f4e5d49","status":"stopped","cluster":"jm","indexerImage":"onfinality/subql-node:v0.28.2","queryImage":"onfinality/subql-query:v0.12.0","subFolder":"subql/ethereum-v1/pangolin","endpoint":null,"dictEndpoint":"","type":"stage","queryUrl":"https://api.subquery.network/sq/fewensa/subql-bridge-pangolin__ZmV3Z","configuration":{"config":{"query":{},"indexer":{"batchSize":30}},"role":"master"}}]
Error: 
   0: unknown variant `stopped`, expected one of `processing`, `running`, `error` at line 1 column 207

```

**Comment by fewensa** [comment]
Please don't merge until I have fully tested

**Comment by HackFisher** [comment]
ready for review and merge?

**Comment by fewensa** [comment]
> 

Yes, please let me check the subql deployment

File: frame/assistants/shadow-liketh/src/shadow.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            member_leaf_index,
+            last_leaf_index: last_leaf_index - 1,
+            proof,
+        };
+        Ok(EthereumReceiptWithMMRProof { receipt, mmr_proof })
+    }
+}
+
+impl Shadow {
+    fn extract_peaks(&self, positions: Vec<MMRNode>) -> Vec<(u64, [u8; 32])> {
```
Comment (HackFisher) [inline frame/assistants/shadow-liketh/src/shadow.rs:119]:
```suggestion
    fn extract_positions_and_hashs(&self, positions: Vec<MMRNode>) -> Vec<(u64, [u8; 32])> {
```

**Review by HackFisher — APPROVED** [review]

**Comment by fewensa** [comment]
All subqls deployment check done.

File: frame/assistants/shadow-liketh/src/shadow.rs (hunk 10 above)
Comment (fewensa) [reply frame/assistants/shadow-liketh/src/shadow.rs]: Please wait,. let me update reference after this chagned
